source/h2_stream.c

```diff
@@ -0,0 +1,434 @@
+        (void *)(stream),                                                                                              \
+        aws_h2_stream_state_to_str((stream)->state),                                                                   \
+        __VA_ARGS__)
+#define STREAM_LOG(level, stream, text) STREAM_LOGF(level, stream, "%s", text)
+
+static void s_stream_destroy(struct aws_http_stream *stream_base);
+
+struct aws_http_stream_vtable s_h2_stream_vtable = {
+    .destroy = s_stream_destroy,
+    .update_window = NULL,
+};
+
+/* Shortcut for logging invalid stream and raising an error */
+static int s_h2_stream_raise_invalid_frame(struct aws_h2_stream *stream, enum aws_h2_frame_type type, int error_code) {
+    STREAM_LOGF(
+        ERROR,
+        stream,
+        "Not allowed to receive frame of type %s when in %s state, raising %s",
+        aws_h2_frame_type_to_str(type),
+        aws_h2_stream_state_to_str(stream->state),
+        aws_error_name(error_code));
+    return aws_raise_error(error_code);
+}
+
+static void s_h2_stream_set_state(struct aws_h2_stream *stream, enum aws_h2_stream_state new_state) {
+    STREAM_LOGF(
+        DEBUG,
+        stream,
+        "Stream moving from state %s to %s",
+        aws_h2_stream_state_to_str(stream->state),
+        aws_h2_stream_state_to_str(new_state));
+
+    stream->state = new_state;
+}
+
+const char *aws_h2_stream_state_to_str(enum aws_h2_stream_state state) {
+    switch (state) {
+        case AWS_H2_STREAM_STATE_IDLE:
+            return "IDLE";
+        case AWS_H2_STREAM_STATE_RESERVED_LOCAL:
+            return "RESERVED_LOCAL";
+        case AWS_H2_STREAM_STATE_RESERVED_REMOTE:
+            return "RESERVED_REMOTE";
+        case AWS_H2_STREAM_STATE_OPEN:
+            return "OPEN";
+        case AWS_H2_STREAM_STATE_HALF_CLOSED_LOCAL:
+            return "HALF_CLOSED_LOCAL";
+        case AWS_H2_STREAM_STATE_HALF_CLOSED_REMOTE:
+            return "HALF_CLOSED_REMOTE";
+        case AWS_H2_STREAM_STATE_CLOSED:
+            return "CLOSED";
+        default:
+            /* unreachable */
+            AWS_FATAL_ASSERT(false && "Unknown value for aws_h2_stream_state");
+            return NULL;
+    }
+}
+
+/***********************************************************************************************************************
+ * Frame Handling
+ **********************************************************************************************************************/
+static int s_h2_stream_handle_data(struct aws_h2_stream *stream, struct aws_h2_frame_decoder *decoder) {
+    AWS_PRECONDITION(decoder->header.type == AWS_H2_FRAME_T_DATA);
+
+    struct aws_h2_frame_data frame;
+    if (aws_h2_frame_data_decode(&frame, decoder)) {
+        return AWS_OP_ERR;
+    }
+
+    /* Call user callback */
+    stream->base.on_incoming_body(&stream->base, &frame.data, stream->base.user_data);
+
+    /* Send window increment packet */
+    struct aws_h2_frame_window_update window_update;
```

Comment:
on_incoming_body doesn't have the weird window_update_out_param thing anymore.
Not sure you want the window_frame_update_encoding logic in this body-decoding function